File: js/test/otp.test.ts
import { describe, it, expect } from "vitest";
import { CallistoClient } from "../src/client.js";
import { OtpProvider } from "../src/models.js";
import { mockFetch } from "./helpers.js";

const opts = { clientId: "cid", apiKey: "k", baseUrl: "https://api.test/v1" };

describe("otp", () => {
  it("send POSTs /otp/send", async () => {
    const { fetch, captured } = mockFetch(200, {
      id: "o1", provider: "sms", recipient: {}, expires_at: "x", expires_in: 300,
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.otp.send({ to: "2250700000000", message: "Code {code}" });
    expect(captured[0].url).toBe("https://api.test/v1/otp/send");
    expect(captured[0].body).toEqual({ to: "2250700000000", message: "Code {code}" });
    expect(res.id).toBe("o1");
  });

  it("rejects whatsapp provider without instance_code before hitting the network", async () => {
    const { fetch, captured } = mockFetch(200, {});
    const client = new CallistoClient(opts, fetch);
    await expect(
      client.otp.send({ to: "x", message: "m", provider: OtpProvider.Whatsapp }),
    ).rejects.toThrow(/instance_code/);
    expect(captured.length).toBe(0);
  });

  it("sends instance_code on the wire as instanceCode for whatsapp", async () => {
    const { fetch, captured } = mockFetch(200, {
      id: "o1", provider: "whatsapp", recipient: {}, expires_at: "x", expires_in: 300,
    });
    const client = new CallistoClient(opts, fetch);
    await client.otp.send({
      to: "2250700000000",
      message: "Code {code}",
      provider: OtpProvider.Whatsapp,
      instance_code: "inst_1",
    });
    const body = captured[0].body as Record<string, unknown>;
    expect(body.instanceCode).toBe("inst_1");
    expect(body.instance_code).toBeUndefined();
  });

  it("verify POSTs /otp/verify", async () => {
    const { fetch, captured } = mockFetch(200, {
      id: "o1", status: "verified", verified: true, verified_at: "now",
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.otp.verify({ otp_id: "o1", code: "12345" });
    expect(captured[0].url).toBe("https://api.test/v1/otp/verify");
    expect(res.verified).toBe(true);
  });

  it("getStatus GETs /otps/:id", async () => {
    const { fetch, captured } = mockFetch(200, {
      otp_id: "o1", status: "pending", recipient: "+2250700000000",
      expires_at: "2026-06-01 10:05:00", verified_at: null,
      attempts: 0, created_at: "2026-06-01 10:00:00",
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.otp.getStatus("o1");
    expect(captured[0].url).toBe("https://api.test/v1/otps/o1");
    expect(res.otp_id).toBe("o1");
    expect(res.status).toBe("pending");
  });

  it("list GETs /otps", async () => {
    const { fetch, captured } = mockFetch(200, {
      items: [{
        id: "o1", status: "verified", recipient: "+2250700000000",
        expires_at: "2026-06-01 10:05:00", verified_at: "2026-06-01 10:02:00",
        attempts: 1, created_at: "2026-06-01 10:00:00",
      }],
      total: 1, per_page: 20, current_page: 1, next: 1, previous: 1, total_pages: 1,
    });
    const client = new CallistoClient(opts, fetch);
    const res = await client.otp.list({ page: 1, limit: 10 });
    expect(captured[0].url).toBe("https://api.test/v1/otps?page=1&limit=10");
    expect(res.total).toBe(1);
    expect(res.items[0].id).toBe("o1");
    expect(res.items[0].status).toBe("verified");
  });
});

File: java/src/test/java/com/callisto/sdk/OtpTest.java
package com.callisto.sdk;

import com.callisto.sdk.enums.OtpProvider;
import com.callisto.sdk.enums.OtpType;
import com.callisto.sdk.errors.ValidationException;
import com.callisto.sdk.models.SendOtpResult;
import com.callisto.sdk.models.VerifyOtpResult;
import com.callisto.sdk.resources.OtpSendRequest;
import com.fasterxml.jackson.databind.JsonNode;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class OtpTest {

    private static final String SEND_RESPONSE =
            "{\"id\": \"otp_1\", \"provider\": \"sms\", \"recipient\": {}, "
                    + "\"expires_at\": \"2026\", \"expires_in\": 300}";

    @Test
    void sendBuildsBodyWithEnumValues() {
        StubHttpClient stub = new StubHttpClient(200, SEND_RESPONSE);
        try (CallistoClient client = TestSupport.client(stub)) {
            SendOtpResult result = client.otp().send(OtpSendRequest.builder()
                    .to("+225070").message("Code {code}")
                    .type(OtpType.DIGIT).digitSize(6).expiredIn(300).build());

            assertEquals("POST", stub.method());
            assertEquals("https://api.example.com/v1/otp/send", stub.uri().toString());
            JsonNode body = TestSupport.json(stub.lastBody);
            assertEquals("digit", body.get("type").asText());
            assertEquals(6, body.get("digit_size").asInt());
            assertEquals(300, body.get("expired_in").asInt());
            assertEquals("otp_1", result.getId());
        }
    }

    @Test
    void sendWhatsappRequiresInstanceCode() {
        StubHttpClient stub = new StubHttpClient(200, SEND_RESPONSE);
        try (CallistoClient client = TestSupport.client(stub)) {
            assertThrows(ValidationException.class, () -> client.otp().send(
                    OtpSendRequest.builder().to("+1").message("m")
                            .provider(OtpProvider.WHATSAPP).build()));
        }
    }

    @Test
    void sendWhatsappSerializesInstanceCodeKey() {
        StubHttpClient stub = new StubHttpClient(200, SEND_RESPONSE);
        try (CallistoClient client = TestSupport.client(stub)) {
            client.otp().send(OtpSendRequest.builder().to("+1").message("m")
                    .provider("whatsapp").instanceCode("inst_1").build());
            JsonNode body = TestSupport.json(stub.lastBody);
            assertEquals("whatsapp", body.get("provider").asText());
            assertEquals("inst_1", body.get("instanceCode").asText());
        }
    }

    @Test
    void verifyPostsBody() {
        StubHttpClient stub = new StubHttpClient(200,
                "{\"id\": \"otp_1\", \"status\": \"verified\", \"verified\": true}");
        try (CallistoClient client = TestSupport.client(stub)) {
            VerifyOtpResult result = client.otp().verify("otp_1", "123456");
            assertEquals("https://api.example.com/v1/otp/verify", stub.uri().toString());
            JsonNode body = TestSupport.json(stub.lastBody);
            assertEquals("otp_1", body.get("otp_id").asText());
            assertEquals("123456", body.get("code").asText());
            assertTrue(result.isVerified());
        }
    }

    @Test
    void getStatusUsesOtpsPath() {
        StubHttpClient stub = new StubHttpClient(200, "{\"otp_id\": \"otp_1\", \"status\": \"pending\"}");
        try (CallistoClient client = TestSupport.client(stub)) {
            client.otp().getStatus("otp_1");
            assertEquals("https://api.example.com/v1/otps/otp_1", stub.uri().toString());
        }
    }

    @Test
    void listUsesLimitQueryKey() {
        StubHttpClient stub = new StubHttpClient(200,
                "{\"items\": [], \"total\": 0, \"per_page\": 0, \"current_page\": 1, "
                        + "\"next\": null, \"previous\": null, \"total_pages\": 0}");
        try (CallistoClient client = TestSupport.client(stub)) {
            client.otp().list(null, null, 1, 20);
            String uri = stub.uri().toString();
            assertTrue(uri.contains("/otps?"), uri);
            assertTrue(uri.contains("limit=20"), uri);
            assertFalse(uri.contains("per_page"), uri);
        }
    }
}
